### PR TITLE
Cleanup extensibility

### PR DESCRIPTION
**Pull request by kumare3**
Implement 2 extensions
Generic SQL task base and extended Spark Task

### CODE REVIEW

File: flytekit/annotated/workflow.py [hunk 1]
```diff
@@ -115,14 +116,15 @@ def wrapper(fn):
...
         output_names = list(interface.outputs.keys())
-        for i, out in enumerate(workflow_outputs):
-            output_name = output_names[i]
-            # TODO: Check that the outputs returned type match the interface.
-            # output_literal_type = out.literal_type
-            # logger.debug(f"Got output wrapper: {out}")
-            # logger.debug(f"Var name {output_name} wf output name {outputs[i]} type: {output_literal_type}")
-            binding_data = _literal_models.BindingData(promise=out)
-            bindings.append(_literal_models.Binding(var=output_name, binding=binding_data))
+        if len(output_names) > 0:
```
Comment (wild-endeavor) [inline flytekit/annotated/workflow.py:119]: Why this change?  Do we not trust `workflow_outputs`?

File: flytekit/annotated/task.py [hunk 2]
```diff
@@ -1,18 +1,20 @@
...
 from flytekit.common import nodes as _nodes, interface as _common_interface
 from flytekit.common.exceptions import user as _user_exceptions
 from flytekit.common.promise import NodeOutput as _NodeOutput
 from flytekit.models import task as _task_model, literals as _literal_models
 from flytekit.models.core import workflow as _workflow_model, identifier as _identifier_model
-import inspect
-from flytekit.annotated.interface import transform_signature_to_typed_interface
 
 
 # This is the least abstract task. It will have access to the loaded Python function
```
Comment (wild-endeavor) [inline flytekit/annotated/task.py:20]: Is this comment correct?  Do you mean, "most abstract task"?

File: flytekit/annotated/task.py [hunk 3]
```diff
@@ -155,53 +184,82 @@ def name(self) -> str:
 
 class PythonFunctionTask(Task):
 
-    def __init__(self, task_function: Callable, metadata: _task_model.TaskMetadata, *args, **kwargs):
-        interface = transform_signature_to_typed_interface(inspect.signature(task_function))
+    def __init__(self, task_function: Callable, metadata: _task_model.TaskMetadata, ignore_input_vars: List[str] = None,
```
Comment (wild-endeavor) [inline flytekit/annotated/task.py:187]: What is ignore_input_vars for?  Can we document the use case?

File: flytekit/annotated/interface.py [hunk 4]
```diff
@@ -1,14 +1,89 @@
...
+from typing import Dict, Generator, Union, Type, Tuple, List
 
 from flytekit import logger
 from flytekit.annotated import type_engine
 from flytekit.common import interface as _common_interface
 from flytekit.models import interface as _interface_models
 
 
-def transform_signature_to_typed_interface(signature: inspect.Signature) -> _common_interface.TypedInterface:
+class Interface(object):
```
Comment (wild-endeavor) [inline flytekit/annotated/interface.py:12]: Can we talk about this next week?

File: flytekit/annotated/task.py [hunk 5]
```diff
@@ -155,53 +184,82 @@ def name(self) -> str:
...
-        return outputs_literal_map
+    def native_interface(self) -> Interface:
+        return self._native_interface
 
-    def _passthrough_execute(self, **kwargs):
+
+class PysparkFunctionTask(PythonFunctionTask):
+    def __init__(self, task_function: Callable, metadata: _task_model.TaskMetadata, *args, **kwargs):
+        super(PysparkFunctionTask, self).__init__(task_function, metadata,
+                                                  ignore_input_vars=["spark_session", "spark_context"], *args, **kwargs)
```
Comment (wild-endeavor) [inline flytekit/annotated/task.py:205]: I feel like I'd rather add more stuff to `PysparkFunctionTask` and subclass `Task` directly, instead of adding functionality at the base layer to remove/add inputs.  Or do you foresee other tasks needing this?

Should spark_session and spark_context be valid inputs for data catalog?

File: flytekit/annotated/task.py [hunk 6]
```diff
@@ -80,13 +82,41 @@ def _create_and_link_node(self, ctx: FlyteContext, *args, **kwargs):
...
+        else:
+            # Question: How do you know you're going to enumerate them in the correct order? Even if autonamed, will
+            # output2 come before output100 if there's a hundred outputs? We don't! We'll have to circle back to
+            # the Python task instance and inspect annotations again. Or we change the Python model representation
+            # of the interface to be an ordered dict and we fill it in correctly to begin with.
+            native_outputs_as_map = {expected_output_names[i]: native_outputs[i] for i, _ in
+                                     enumerate(native_outputs)}
+
+        # We manually construct a LiteralMap here because task inputs and outputs actually violate the assumption
+        # built into the IDL that all the values of a literal map are of the same type.
```
Comment (katrogan) [inline flytekit/annotated/task.py:109]: can we change idl then?

File: flytekit/annotated/task.py [hunk 7]
```diff
@@ -80,13 +82,41 @@ def _create_and_link_node(self, ctx: FlyteContext, *args, **kwargs):
...
+        outputs_literal_map = _literal_models.LiteralMap(literals={
+            k: flytekit_engine.python_value_to_idl_literal(ctx, v, self.interface.outputs[k].type) for k, v in
+            native_outputs_as_map.items()
+        })
+        print("Outputs!")
```
Comment (katrogan) [inline flytekit/annotated/task.py:114]: nit: rm?
